### DISCUSSION TITLE
[next.config.js] rewrites be able to change the proxyTimeout

### DISCUSSION THREAD

**Discussion by viniciusbitt** [vercel/next.js#36598] [Ideas]
### Describe the feature you'd like to request

I am creating a site in Next.JS and my server in Express and I am doing rewrite inside `next.config.js` pointing to my server.

```
const nextConfig = {
  async rewrites() {
    return [
      {
        source: "/api/:path*",
        destination: `${server.url}api/:path*`
      },
    ];
  },
};
```

But there's a problem, I can't make a request to my server and the response takes longer than `30 seconds`, because the following error appears:

`Error: Socket hang up`

Also happens if I am using for example Socket.IO, where this error appears in the console.

### Describe the solution you'd like

Checking where the possible problem is, I realized it is on this lines:

https://github.com/vercel/next.js/blob/0de109baab7dbed248cc9b5dbc2ea85535c8580d/packages/next/server/next-server.ts#L486-L492

where `proxyTimeout` is being limited to `30 seconds`.

Thinking about a solution, I believe it should be possible to configure `proxyTimeout`.

Related issues:
#36251

Related discussions:
#24805

### Describe alternatives you've considered

N/A

**Comment by BahaaZidan** [comment]
Keeping this active. We really need this feature.

**Comment by liamlows** [comment]
Does anyone know if we will see this feature added soon? We have a use case where an executable can only be downloaded if a JWT is authorized and verified thus we cant use a S3 bucket or equivalent and it needs to go through an API. 

If we use the next API it yells at us over the request size saying its too big, if we use the rewrites feature it times out and throws a socket hangup. 

We really can't win with our current options so this feature would help a lot...

**Comment by ayukuzmenko** [comment]
This feature already added. You can use an experimental option **proxyTimeout**. Pull request [#40289](https://github.com/vercel/next.js/pull/40289).

**Comment by QingjiaTsang** [comment]
proxyTimeout is a really important feature when my proxy API gets a long time to respond

**Comment by Osama-V2** [comment]
Just type proxyTimeout: in your vsCode search and make sure you're searching in node_module and change 
proxyTimeout: proxyTimeout === null ? undefined : proxyTimeout || 30000
to
proxyTimeout: proxyTimeout === null ? undefined : proxyTimeout || 300000
or any value from your choice.

:)

**Reply by QingjiaTsang** [reply]
No need, dude. Just go get next.config.ts and set the proxyTimeout with any number value in the experimental option.

**Reply by Osama-V2** [reply]
lol i didn't know that, thanks tho <3

**Reply by hassanpasha5630** [reply]
Do you have an example of how to do this ?

**Reply by VettedReact22** [reply]
Can you please tell me in next-server.js where exactly to write this line proxyTimeout: proxyTimeout === null ? undefined : proxyTimeout || 300000

**Reply by Osama-D** [reply]
Honestly i did it and it did not work, just search in vsCode search bar and make sure you're searching in node_module and change the value from 30000 to higher value like 3000000000

**Reply by hassanpasha5630** [reply]
This solved my issue this is how I did it added this at the root vercel.json 

note I'm on their pro plan  @VettedReact22 

{
  "functions": {
    "api/index.py": {
      "memory": 3008,
      "maxDuration": 300
    },
    "api/*.py": {
      "memory": 3008,
      "maxDuration": 300
    }
  }
}

@@

**Reply by QingjiaTsang** [reply]
```
    experimental: {
      proxyTimeout: 1000 * 120,
    },
```

Hey guys, I configured the code above at next.config.js before on nextjs13.4 but now I've upgraded to nextjs14. I press Ctrl and click the experimental and it won't bring me to the type file which I remember it would do before and some experimental configurations would be shown there. So IMO, maybe nextjs change something on this config.

**Reply by AdolfVonKleist** [reply]
@QingjiaTsang thank you!  You ended 2 hrs of fruitless headache!

**Reply by bexorg-gsurtani** [reply]
This was the fix.  Thank you

**Reply by Dev-Andejecruher** [reply]
Muchas Gracias por el aporte, me sirvió de mucho.

**Reply by thucne** [reply]
> ```
>     experimental: {
>       proxyTimeout: 1000 * 120,
>     },
> ```
> 
> Hey guys, I configured the code above at next.config.js before on nextjs13.4 but now I've upgraded to nextjs14. I press Ctrl and click the experimental and it won't bring me to the type file which I remember it would do before and some experimental configurations would be shown there. So IMO, maybe nextjs change something on this config.

Thank you, I've been searching for this this entire afternoon, extremely helpful, still, the limit to have proxied requests is so painful...

**Comment by wiiiimm** [comment]
Vercel [currently limits proxy requests to 30 seconds](https://vercel.com/docs/limits/overview#proxied-request-timeout).

> The amount of time (in seconds) that a proxied request (rewrites or routes with an external destination) is allowed to process an HTTP request. The maximum timeout is 30 seconds. If the external server does not reply until the maximum timeout is reached, an error with the message ROUTER_EXTERNAL_TARGET_ERROR will be returned.

So until this limit is changed, the `proxyTimeout` in Next.js can only be anything up to 30 seconds if it is to be hosted on Vercel.

**Reply by thucne** [reply]
Update: this limit has been increased to 120 seconds 😊
https://vercel.com/docs/limits#proxied-request-timeout
